File: samples/go/aggregation_svc/svc/svc.go
/*
 * Copyright 2022 CECTC, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package svc

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"

	"github.com/dbpack/samples/order_svc/dao"
	dao2 "github.com/dbpack/samples/product_svc/dao"
)

type Svc struct {
}

func GetSvc() *Svc {
	return &Svc{}
}

func (svc *Svc) CreateSo(ctx context.Context, xid string, rollback bool) error {
	soMasters := []*dao.SoMaster{
		{
			BuyerUserSysNo:       10001,
			SellerCompanyCode:    "SC001",
			ReceiveDivisionSysNo: 110105,
			ReceiveAddress:       "beijing",
			ReceiveZip:           "000001",
			ReceiveContact:       "scott",
			ReceiveContactPhone:  "18728828296",
			StockSysNo:           1,
			PaymentType:          1,
			SoAmt:                6999 * 2,
			Status:               10,
			AppID:                "dk-order",
			SoItems: []*dao.SoItem{
				{
					ProductSysNo:  1,
					ProductName:   "apple iphone 13",
					CostPrice:     6799,
					OriginalPrice: 6799,
					DealPrice:     6999,
					Quantity:      2,
				},
			},
		},
	}

	reqs := []*dao2.AllocateInventoryReq{{
		ProductSysNo: 1,
		Qty:          2,
	}}

	type rq1 struct {
		Req []*dao.SoMaster
	}

	type rq2 struct {
		Req []*dao2.AllocateInventoryReq
	}

	q1 := &rq1{Req: soMasters}
	soReq, err := json.Marshal(q1)
	req1, err := http.NewRequest("POST", "http://localhost:3001/createSo", bytes.NewBuffer(soReq))
	if err != nil {
		panic(err)
	}
	req1.Header.Set("Content-Type", "application/json")
	req1.Header.Set("xid", xid)

	client := &http.Client{}
	result1, err1 := client.Do(req1)
	if err1 != nil {
		return err1
	}

	if result1.StatusCode == 400 {
		return errors.New("err")
	}

	q2 := &rq2{
		Req: reqs,
	}
	ivtReq, _ := json.Marshal(q2)
	req2, err := http.NewRequest("POST", "http://localhost:3002/allocateInventory", bytes.NewBuffer(ivtReq))
	if err != nil {
		panic(err)
	}
	req2.Header.Set("Content-Type", "application/json")
	req2.Header.Set("xid", xid)

	result2, err2 := client.Do(req2)
	if err2 != nil {
		return err2
	}

	if result2.StatusCode == 400 {
		return errors.New("err")
	}

	if rollback {
		fmt.Println("rollback")
		return errors.New("there is a error")
	}
	return nil
}

File: samples/php/aggregation_svc/svc/svc.php
<?php

class AggregationSvc
{

    public function CreateSo(string $xid, bool $rollback): bool
    {
        $createSoSuccess = $this->createSoRequest($xid);
        if (!$createSoSuccess) {
            return false;
        }
        $allocateInventorySuccess = $this->allocateInventoryRequest($xid);
        if (!$allocateInventorySuccess) {
            return false;
        }
        if ($rollback) {
            return false;
        }
        return true;
    }

    protected function createSoRequest(string $xid): bool
    {
        $soMasters = [
            [
                'buyerUserSysNo' => 10001,
                'sellerCompanyCode' => 'SC001',
                'receiveDivisionSysNo' => 110105,
                'receiveAddress' => 'beijing',
                'receiveZip' => '000001',
                'receiveContact' => 'scott',
                'receiveContactPhone' => '18728828296',
                'stockSysNo' => 1,
                'paymentType' => 1,
                'soAmt' => 6999 * 2,
                'status' => 10,
                'appID' => 'dk-order',
                'soItems' => [
                    [
                        'productSysNo' => 1,
                        'productName' => "apple iphone 13",
                        'costPrice' => 6799,
                        'originalPrice' => 6799,
                        'dealPrice' => 6999,
                        'quantity' => 2,
                    ],
                ],
            ],
        ];
        $url = 'http://localhost:3001/createSo';
        $response = $this->sendRequest($url, $xid, $soMasters);
        if ($response === false) {
            return false;
        }
        $responseData = json_decode($response, true);
        return $responseData['success'] ?? false;
    }

    protected function allocateInventoryRequest(string $xid): bool
    {
        $allocateInventoryReq = [
            [
                'ProductSysNo' => 1,
                'Qty' => 2,
            ],
        ];
        $url = 'http://localhost:3002/allocateInventory';
        $response = $this->sendRequest($url, $xid, $allocateInventoryReq);
        if ($response === false) {
            return false;
        }
        $responseData = json_decode($response, true);
        return $responseData['success'] ?? false;
    }

    private function sendRequest(string $url, string $xid, array $data): bool|string
    {
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => array(
                'Content-Type: application/json',
                'xid: ' . $xid,
            ),
            CURLOPT_POSTFIELDS => json_encode($data)
        ]);

        $response = curl_exec($ch);
        if ($response === false) {
            return false;
        }
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        if ($httpCode == 400) {
            return false;
        }
        curl_close($ch);

        return $response;
    }
}